### ISSUE TITLE
Use after free / double free when using memoryviews in python 2

### ISSUE THREAD

**Issue by nostrad** [pybind/pybind11#973]
## Issue description

When nesting uses of memoryviews in python 2 (2.17.3 on Debian Stretch), buffer_info will get free'd twice.

The below code is always reproducible, removing the t=memoryview(t2) wrapping makes it work once again. Looking at numpy (https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/buffer.c#L85) it seems that bf_releasebuffer is unsafe on python 2.

## Reproducible example code

test_buf.py:
```
import test
import struct

t2 = test.test()
t = memoryview(t2) # This fails
# t = t2 # This works
struct.unpack_from("10B", t)
# x = memoryview(t[2:3]) # Using this instead of struct.unpack will also trigger the error
```

test.cc:
```c++
#include <iostream>
#include <sstream>
#include <string>
#include <pybind11/pybind11.h>

namespace py = pybind11;

class Test {
public:
        Test() : d(10*sizeof(int), 0) {}
        std::vector<unsigned char> d;
};

PYBIND11_PLUGIN(test)
{
        py::module m("test");
        py::class_<Test>(m, "test", py::buffer_protocol())
                .def(py::init<>())
                .def_buffer([](Test &t) -> py::buffer_info {
                        return py::buffer_info(t.d.data(),
                                               sizeof(unsigned char),
                                               py::format_descriptor<unsigned char>::format(),
                                               1,
                                               { t.d.size() },
                                               { sizeof(unsigned char) });
                           });
        return m.ptr();
}
```

Makefile:
```
PY=python2
ASAN=n

CFLAGS-$(ASAN) += -fsanitize=address
PYENV-$(ASAN) += 'LD_PRELOAD="/usr/lib/x86_64-linux-gnu/libasan.so.3"'

run: test.so
        $(PYENV-y) $(PY) test_buf.py

test.so:
        g++ $(CFLAGS-y) -fPIC -shared -Wall -std=c++14 $(shell pkg-config --libs $(PY2)) $(shell pkg-config --cflags $(PY)) -o test.so test.cc

clean:
        rm test.so
```

**Comment by dean0x7d** [comment]
Thanks for the reproducible example! I've been able to confirm the bug and it looks like the underlying issue is in Python 2.7's `memoryview`. Unfortunately, I don't see a way to work around it on our end.

As far as I've been able to gather, `bf_releasebuffer` itself is fine and this is a bug with the `memoryview` implementation in Python 2.7 and early 3.x. These seem to be the relevant CPython issues: [7433](https://bugs.python.org/issue7433), [9990](https://bugs.python.org/issue9990), [10181](https://bugs.python.org/issue10181). The fix was submitted [here](https://mail.python.org/pipermail/python-checkins/2012-February/111343.html) but only for Python >= 3.3. From the discussion in [10181](https://bugs.python.org/issue10181), it looks like a fix for 2.7 was considered but dropped.

**Comment by nostrad** [comment]
Pretty much what I expected then.

The current hope is that our codebase will jump to python3 quite soon, until then I have kludged around it locally by hacking in my own implementation for the one bf_getbuffer where I need the functionality.
